bus-mapping/src/circuit_input_builder/input_state_ref.rs

```diff
@@ -1549,88 +1600,459 @@ impl<'a> CircuitInputStateRef<'a> {
         dst_addr: u64,
         src_addr_end: u64,
         bytes_left: u64,
-    ) -> Result<Vec<(u8, bool)>, Error> {
+    ) -> Result<Vec<(u8, bool, bool)>, Error> {
         let mut copy_steps = Vec::with_capacity(bytes_left as usize);
-        for idx in 0..bytes_left {
-            let addr = src_addr.checked_add(idx).unwrap_or(src_addr_end);
-            let step = if addr < src_addr_end {
-                let code = bytecode.code.get(addr as usize).unwrap();
-                (code.value, code.is_code)
+        if bytes_left == 0 {
+            return Ok(copy_steps);
+        }
+
+        let (_, dst_begin_slot) = self.get_addr_shift_slot(dst_addr).unwrap();
+        let (_, dst_end_slot) = self.get_addr_shift_slot(dst_addr + bytes_left).unwrap();
+        let mut memory = self.call_ctx_mut()?.memory.clone();
```

Comment:
This seems to read the memory before the copy. Shouldn’t it read from the bytecode like before?

Reply:
the byte code has been copied into  ` let mut memory = self.call_ctx_mut()?.memory.clone();`. in its `gen_associated_ops` in buss mapping (e.g.  codecopy:  `memory.copy_from(dst_offset, code_offset, length, &code)` ), so that we can retrieve it from memory here.